docs/guide/editing-custom-widgets-in-context.md

```diff
@@ -0,0 +1,288 @@
+
+The `defaultData` option of the widget module may be set to a default value for newly created widgets of this type.
+
+### The widget editor
+
+The Vue component name configured via the `widgetEditor` subproperty of the `components` option (the "widget editor") must display the content of the widget in an editable form. To provide a contextual editing experience, that interface should be similar to the ordinary read-only view generated by `widget.html`.
+
+The widget editor receives a `value` Vue prop containing the existing value of the widget (an object), and emits an `update` Vue event with a new object whenever appropriate. The component should *not* modify the `value` prop given to it.
+
+The widgetEditor **must not** attempt to save changes by itself. Instead it must emit the `update` event and let Apostrophe takes care of the rest. Never assume the widget will be in a particular document type. It may be in any area, nested in any document type. Apostrophe will handle this for you.
+
+### Debouncing update events
+
+If the value will be changing quickly, for instance as the user types, performance can suffer if an `update` event is generated for every keystroke. The widget editor can speed up the interface by emitting an `update` event no more than once per second. This is called "debouncing."
+
+As a hint that input might not yet be saved, a widget that "debounces" `update` events should emit a `context-editing` event on every change, even if `update` is intentionally delayed. However this should only be done if the `docId` prop passed to the widget editor is equal to `windows.apos.adminBar.contextId`.
+
+If the `focused` prop of the widget editor becomes `false` and an `update` event has been delayed for debouncing purposes, the widget should cancel its timer and immediately emit the update. Use Vue's `watch` feature to monitor `focused`.
+
+### Saving the content
+
+By default the `fields` section of a contextually edited widget has no effect on the user interface. Emitting a reasonable value is the task of the custom widget editor component. However configuring fields is still necessary since the server will still use them to sanitize the data before saving it.
+
+If no `fields` section is configured, no data will be saved at all unless the `sanitize` method of the module is overridden to provide an alternative way to verify user input, as is done in the case of the core rich text widget.
+
+## Example
+
+Here is a simple "hero" heading widget. Both the heading and the width of the heading can be adjusted directly on the page. This gives the user immediate visual feedback.
+
+<AposCodeBlock>
+```js
+module.exports = {
+  extend: '@apostrophecms/widget-type',
+  options: {
+    contextual: true,
+    defaultData: {
+      heading: '',
+      width: 100
+    },
+    components: {
+      widgetEditor: 'HeroWidgetEditor'
+    }
```

Comment:
If all these are underpinned by the widget being contextual should the config look like this?
```
options: {
  contextual: {
    defaultData: { .. },
    components: {
      widgetEditor: 'HeroWidgetEditor'
    } 
  }
}
```

Reply:
`widgetEditor` is applicable to non-contextual widgets too, if you want to swap out the schema-based dialog editor for whatever reason, and also I'm documenting what exists (and what we have to maintain bc with at this point), so I think it's best not to change this.